codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/ServerCodegenVisitor.kt

```diff
@@ -166,12 +184,53 @@ class ServerCodegenVisitor(context: PluginContext, private val codegenDecorator:
         logger.info("[rust-server-codegen] Generating a structure $shape")
         rustCrate.useShapeWriter(shape) { writer ->
             StructureGenerator(model, symbolProvider, writer, shape).render(CodegenTarget.SERVER)
-            val builderGenerator =
-                BuilderGenerator(codegenContext.model, codegenContext.symbolProvider, shape)
+            val builderGenerator = ServerBuilderGenerator(
+                codegenContext,
+                shape,
+                takeInUnconstrainedTypes = shapesReachableFromOperationInputs.contains(shape)
+            )
             builderGenerator.render(writer)
             writer.implBlock(shape, symbolProvider) {
                 builderGenerator.renderConvenienceMethod(this)
             }
+
+            if (shapesReachableFromOperationInputs.contains(shape)) {
+                ServerStructureConstrainedTraitImpl(symbolProvider, shape, writer).render()
+            }
+        }
+    }
+
+    override fun listShape(shape: ListShape) {
+        if (shapesReachableFromOperationInputs.contains(shape) && shape.canReachConstrainedShape(model, symbolProvider)) {
+            logger.info("[rust-server-codegen] Generating an unconstrained type for list $shape")
+
+            rustCrate.withModule(unconstrainedModule) { writer ->
+                UnconstrainedListGenerator(
+                    model,
+                    symbolProvider,
+                    unconstrainedShapeSymbolProvider,
+                    constraintViolationSymbolProvider,
+                    writer,
+                    shape
+                ).render()
+            }
+        }
+    }
+
+    override fun mapShape(shape: MapShape) {
+        if (shapesReachableFromOperationInputs.contains(shape) && shape.canReachConstrainedShape(model, symbolProvider)) {
+            logger.info("[rust-server-codegen] Generating an unconstrained type for map $shape")
+
+            rustCrate.withModule(unconstrainedModule) { writer ->
+                UnconstrainedMapGenerator(
+                    model,
+                    symbolProvider,
+                    unconstrainedShapeSymbolProvider,
+                    constraintViolationSymbolProvider,
+                    writer,
+                    shape
+                ).render()
+            }
```

Comment:
I am probably missing something here.. Can't there be constrained maps and lists?

Reply:
Yes, but constraint traits are not implemented in this PR. This PR only generates the "unconstrained type" corresponding to a constrained shape.